src/xercesc/validators/common/DFAContentModel.cpp

```diff
@@ -1191,12 +1203,21 @@ void DFAContentModel::buildDFA(ContentSpecNode* const curNode)
                     //  table.
                     //
                     statesToDo[curState] = newSet;
-                    fTransTable[curState] = makeDefStateList();
-                    stateTable->put
-                    (
-                        newSet
-                        , new (fMemoryManager) XMLInteger(curState)
-                    );
+                    try
+                    {
+                        fTransTable[curState] = makeDefStateList();
+                        stateTable->put
+                        (
+                            newSet
+                            , new (fMemoryManager) XMLInteger(curState)
+                        );
+                    }
+                    catch( const OutOfMemoryException& e )
+                    {
+                        oomException = e;
```

Comment:
How safe is the saving and re-throwing of the exception?  Is there any potential for `OutOfMemoryException& e` to be a reference to a base class?  Could `e` end up being truncated as a result?  A direct `throw` to rethrow the existing exception might be safer overall, and avoid the need for a goto.

Could we achieve the same effect with a higher-level try/catch block within the function, and avoid the saving of the exception and the goto?

Reply:
> 

possibly. I didn't want to change the code too much. Your suggestion will probably involve moving the cleanup code in a lambda function that will be called at the end of the nominal code path and in the catch() block.